Fix nstar issues

alfonsoros88 opened the pull request:
This pr addresses the following issues with `nstar`:
- Re-introduces the reconnection loop.
- Fixes some issues with the handling of stdin/stdout that produced some weird artifacts.
- Validates the user input with `clap`
- Add missing help messages for commands

flxo commented on nstar/src/main.rs:
```diff
@@ -55,10 +55,21 @@ async fn send_request(
+        match api::client::Client::new(url).await {
+            Ok(client) => return Ok(client),
+            Err(e) => {
+                time::sleep(time::Duration::from_secs(1)).await;
+                writeln!(output, "{}", e.to_string().red())?;
```
I'd flip the lines and extend the message with a message:

```rust
 writeln!(output, "Failed to connect to {}: {}", url, e.to_string().red())?;
 time::sleep(time::Duration::from_secs(1)).await;
```

flxo commented on nstar/src/main.rs:
```diff
@@ -55,10 +55,21 @@ async fn send_request(
+
+/// Tries to connect to northstar `RECONNECT_RETRIES` times with 1 sec intervals before failing
+async fn try_connect<W: Write>(output: &mut W, url: &url::Url) -> Result<api::client::Client> {
+    for i in 0..RECONNECT_RETRIES {
```
What happens after `RECONNECT_RETRIES`? Wouldn't it make sense to retry forever in interactive mode? 

During development it's convenient to have a console funning forever that connects whenever the rt is reachable instead of having it to restart every while and then.

alfonsoros88 replied:
We can make it infinite, no problem. I just thought that maybe a limit would make sense.

flxo commented on nstar/src/main.rs:
```diff
@@ -71,61 +82,73 @@ async fn print_response<W: Write>(mut out: &mut W, response: Response, json: boo
 
-        writeln!(terminal, "Disconnected")?;
-        time::sleep(time::Duration::from_secs(1)).await;
+    // Shutdown the runtime
+    // Also stops the terminal blocked reading from stdin
+    runtime.shutdown_background();
```
Why is this needed when exiting anyways?

alfonsoros88 replied:
This is needed because the terminal tasks that reads from `stdin` is blocked waiting for input at this point.

flxo replied:
The documentation os `shutdown_backgroud` explains that the tasks are just dropped and leaked. Feels a little bit hacky. Is there a way to unblock the terminal task and terminate gracefully?

alfonsoros88 replied:
I tried a couple of approaches and nothing seems to stop the blocked task. This is the only thing that "worked" so far.

flxo commented on nstar/src/main.rs:
```diff
@@ -71,61 +80,73 @@ async fn print_response<W: Write>(mut out: &mut W, response: Response, json: boo
+                        }
+                    }
+                    input = input.next() => {
+                        let input: &str = if input.is_some() {
+                            input.as_deref().unwrap()
+                        } else {
+                            break 'outer;
+                        };
+                        match parse_prompt(input) {
```
This is easier to read with

```rust
                        let input = if let Some(input) = input {
                            input
                        } else {
                            break 'outer;
                        };
                        match parse_prompt(&input) {
```

alfonsoros88 replied:
Replaced it with:
```rust
let input = input.ok_or_else(|| anyhow!("Failed to read input"))?;
```
It's even nicer ;)

flxo commented on nstar/src/main.rs:
```diff
@@ -55,10 +55,19 @@ async fn send_request(
+    loop {
+        match api::client::Client::new(url)
+            .await
+            .context("Failed to Instantiate client")
```
That a client is instantiated is a impl details that's not interesting for the user. I'd change this text to "Failed to connect to {}"

flxo commented on nstar/src/main.rs:
```diff
@@ -55,10 +55,19 @@ async fn send_request(
+            client => break client,
+        }
+        time::sleep(time::Duration::from_secs(1)).await;
+        writeln!(output, "Retrying for connection")?;
```
"Retrying for connection" feels wrong. "Reconnecting..."?

flxo reviewed (changes requested):
When the rt exits the output is broken:

flxo commented:
@alfonsoros88 Just tested this: When `nstar` is connected and the runtime disconnects, `nstar` doesn't try to reconnect again as we had it before the changes. This is handy when developing northstar or  containers: just keep a `nstar` instance running that automatically reconnects whenever a rt instance is available.

A EOF (ctrl-d) doesn't completely shutdown nstar. This was also the case before the changes.

flxo commented:
The input prompt doesn't work as expected. During the connection attempts there shall be no prompt. In the current state of this PR a prompt is displayed in between the logs about the connection attempt. This can be solved by either dropping there terminal (and making sure the readline task breaks) whenever the nstar is disconnected, or starting the `readline` call in `Terminal` only on request for one input.

```
reconnect...
reconnect...
connected.
unblock readline task -> readline task blocks until eof or input and sends this
readline tasks blocks
...
disconnected...
reconnect...
reconnected...
connected.
unblock readline task etc..
```

alfonsoros88 commented:
after spending a lot of time trying to find a workaround this, I'm not sure what you are asking is even possible with `rustyline` and the way `readline` is implemented. Once blocked on waiting for `stdin` there is not much you can do to interrupt that. I just had to check the "old" version of `nstar` to see how different was it respect to now but it does behave exactly in the same way. That said, I would propose to move forward as it is right now and hopefully we can find some replacement for `rustyline`.

marcmo commented on northstar/src/runtime/minijail.rs:
```diff
@@ -497,24 +497,26 @@ impl Debug {
             .as_ref()
             .and_then(|debug| debug.strace.as_ref())
-        {
-            Some(Strace::new(strace, manifest, &config.log_dir, pid).await?)
-        } else {
-            None
-        };
+            .map(|strace| Strace::new(strace, manifest, &config.log_dir, pid))
+            .into();
```
Didn't know about this OptionFuture...works but does not make the code more readable...or is the behavior changed in any way?

alfonsoros88 replied:
Nope, no difference at all. The reason for this change is to address a complain from clippy nightly that we should use `Option::map` here :man_shrugging:

marcmo commented on nstar/src/main.rs:
```diff
@@ -99,37 +114,39 @@ async fn main() -> Result<()> {
-
         // Try to reconnect
-        client = create_client(&opt.host).await?;
+        writeln!(terminal, "Disconnected, trying to reconnect")?;
+        client = try_connect(&mut terminal, &url).await?;
```
while this is already very good, we could also do a less verbose reconnect by catching failed connects and just adding a `.` to the output

alfonsoros88 replied:
Oh yes, I like this idea :)

marcmo commented:
works much better :) ...I like it

flxo commented:
Hej. Just check this PR. Something seems to be really strange and lock up. After typing some series of commands the prompt is completely messed or locked up somewhere.
![Screenshot 2021-03-22 at 15 37 37](https://user-images.githubusercontent.com/129436/112007210-c4b1c400-8b24-11eb-9500-169197c8ba93.png)
![Screenshot 2021-03-22 at 15 37 51](https://user-images.githubusercontent.com/129436/112007244-cd09ff00-8b24-11eb-89b1-d6cc1fd4c060.png)

![Screenshot 2021-03-22 at 15 43 00](https://user-images.githubusercontent.com/129436/112007873-4f92be80-8b25-11eb-94a4-11886548f820.png)

This happens when the client connected and disconnected from the rt a couple of times. Seems broken.

I think that `rustyline` is the root of all the problems.  Let's discuss the complete removal of rusty line and replace the convenience features with a nice shell completion script that runs `containers` internally to make stateful suggestions.

alfonsoros88 commented:
@flxo it seems that I introduced that error by trying to implement the `Connecting...` message idea. But it seems it's not worth the trouble. I just reverted that change, it should work okay now.

flxo commented:
@alfonsoros88 If there's input while `nstar` is in it's reconnecting loop and it finally reconnects it locks up.

Steps to reproduce:

1. Start runtime
2. Start nstar
3. Stop runtime
4. Input something in nstar and hit return
5. Start the runtime

flxo commented:
Merging this. I'll add a issue to remove the interactive mode.
